lib/ruby_lsp/test_reporter.rb

```diff
@@ -67,6 +67,80 @@ def append_output(message:)
+      # Nil values represent lines for which coverage is not available, like empty lines, comments or keywords like
+      # `else`
+      #
+      # Branches are a hash containing the name of the branch and the location where it is found in tuples with the
+      # following elements: [NAME, ID, START_LINE, START_COLUMN, END_LINE, END_COLUMN] as the keys and the value is the
+      # number of times it was executed
+      #
+      # Methods are a similar hash [ClassName, :method_name, START_LINE, START_COLUMN, END_LINE, END_COLUMN] => NUMBER
+      # OF EXECUTIONS
+      #
+      # Example:
+      # {
+      #   "file_path" => {
+      #     "lines" => [1, 2, 3, nil],
+      #     "branches" => {
+      #       ["&.", 0, 6, 21, 6, 65] => { [:then, 1, 6, 21, 6, 65] => 0, [:else, 5, 7, 0, 7, 87] => 1 }
+      #     },
+      #     "methods" => {
+      #       ["Foo", :bar, 6, 21, 6, 65] => 0
+      #     }
+      # }
+      #: () -> Hash[String, untyped]
+      def gather_coverage_results
+        # Ignore coverage results inside dependencies
+        result = Coverage.result.reject do |file_path, _coverage_info|
+          file_path.start_with?(Bundler.bundle_path.to_s) ||
+            file_path.start_with?(File.dirname(RbConfig::CONFIG["rubylibdir"])) ||
+            file_path.start_with?("eval")
+        end
+
+        result.to_h do |file_path, coverage_info|
+          # Format the branch coverage information as VS Code expects it and then group it based on the start line of
+          # the conditional that causes the branching. We need to match each line coverage data with the branches that
+          # spawn from that line
+          branch_by_line = coverage_info[:branches].flat_map do |branch, data|
+            branch_name = branch[0]
+
+            data.map do |then_or_else, execution_count|
+              name, _id, start_line, start_column, end_line, end_column = then_or_else
+
+              {
+                groupingLine: branch[2],
+                executed: execution_count,
+                location: {
+                  start: { line: start_line, character: start_column },
+                  end: { line: end_line, character: end_column },
+                },
+                label: "#{branch_name} #{name}",
+              }
+            end
+          end.group_by { |branch| branch[:groupingLine] }
+
+          # Format the line coverage information, gathering any branch coverage data associated with that line
+          data = coverage_info[:lines].filter_map.with_index do |execution_count, line_index|
+            next if execution_count.nil?
+
+            {
+              executed: execution_count,
+              location: { line: line_index, character: 0 },
+              branches: branch_by_line[line_index] || [],
+            }
+          end
+
+          # The expected format is URI => { executed: number_of_times_executed, location: { ... }, branches: [ ... ] }
```

Comment:
Can we capture this with a Hash shape in the side instead of `Hash[String, untyped]`?

Reply:
I just figured that if there's going to be a comment to describe the shape, we may as well just formalize it into an RBS comment.

We don't have to describe the shape "all the way down", we can just summarize it to the same level that your comment did, so the method sig above can change like:

```diff
-       #: () -> Hash[String, untyped]
+      #: () -> Hash[String, { executed: Integer, location: Hash[Symbol, untyped], branches: Array }]
```